File: lib_base/src/main/java/com/android/base/architecture/ui/SwipeRefreshView.java
package com.android.base.architecture.ui;


import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

class SwipeRefreshView implements RefreshView {

    private SwipeRefreshLayout mSwipeRefreshLayout;
    private RefreshHandler mRefreshHandler;

    SwipeRefreshView(SwipeRefreshLayout swipeRefreshLayout) {
        mSwipeRefreshLayout = swipeRefreshLayout;
    }

    /**
     * SwipeRefreshLayout
     */
    @Override
    public void autoRefresh() {
        mSwipeRefreshLayout.post(() -> {
            mSwipeRefreshLayout.setRefreshing(true);
            doRefresh();
        });
    }

    @Override
    public void refreshCompleted() {
        mSwipeRefreshLayout.setRefreshing(false);
    }

    @Override
    public void setRefreshHandler(final RefreshHandler refreshHandler) {
        mRefreshHandler = refreshHandler;
        mSwipeRefreshLayout.setOnRefreshListener(this::doRefresh);
    }

    private void doRefresh() {
        if (mRefreshHandler.canRefresh()) {
            mRefreshHandler.onRefresh();
        } else {
            mSwipeRefreshLayout.post(() -> mSwipeRefreshLayout.setRefreshing(false));
        }
    }

    @Override
    public boolean isRefreshing() {
        return mSwipeRefreshLayout.isRefreshing();
    }

    @Override
    public void setRefreshEnable(boolean enable) {
        mSwipeRefreshLayout.setEnabled(enable);
    }

}

File: lib_base/src/main/java/com/android/base/architecture/ui/list/SwipeRefreshView.kt
package com.android.base.architecture.ui.list

import androidx.swiperefreshlayout.widget.SwipeRefreshLayout

internal class SwipeRefreshView(
    private val swipeRefreshLayout: SwipeRefreshLayout
) : RefreshView {

    private var refreshHandler: RefreshView.RefreshHandler? = null

    override fun autoRefresh() {
        swipeRefreshLayout.post {
            swipeRefreshLayout.isRefreshing = true
            doRefresh()
        }
    }

    override fun refreshCompleted() {
        swipeRefreshLayout.isRefreshing = false
    }

    override fun setRefreshHandler(refreshHandler: RefreshView.RefreshHandler?) {
        this.refreshHandler = refreshHandler
        swipeRefreshLayout.setOnRefreshListener { doRefresh() }
    }

    private fun doRefresh() {
        if (refreshHandler?.canRefresh() == true) {
            refreshHandler?.onRefresh()
        } else {
            swipeRefreshLayout.post { swipeRefreshLayout.isRefreshing = false }
        }
    }

    override val isRefreshing: Boolean
        get() = swipeRefreshLayout.isRefreshing

    override var isRefreshEnable: Boolean
        get() = swipeRefreshLayout.isEnabled
        set(enable) {
            swipeRefreshLayout.isEnabled = enable
        }

}